api/v1alpha1/raycluster_webhook.go

```diff
@@ -0,0 +1,310 @@
+		log.Info("setting default network policy client labels", "value", defaultNetworkPolicyClientLabels)
+		r.Spec.NetworkPolicyClientLabels = defaultNetworkPolicyClientLabels
+	}
+	if r.Spec.Worker.Replicas == nil {
+		log.Info("setting default worker replicas", "value", *defaultWorkerReplicas)
+		r.Spec.Worker.Replicas = defaultWorkerReplicas
+	}
+
+	if r.Spec.Image == nil {
+		log.Info("setting default image", "value", *defaultImage)
+		r.Spec.Image = defaultImage
+	} else {
+		if r.Spec.Image.Repository == "" {
+			log.Info("setting default image repository", "value", defaultImage.Repository)
+			r.Spec.Image.Repository = defaultImage.Repository
+		}
+		if r.Spec.Image.Tag == "" {
+			log.Info("setting default image tag", "value", defaultImage.Tag)
+			r.Spec.Image.Tag = defaultImage.Tag
+		}
+	}
+}
+
+//+kubebuilder:webhook:path=/validate-distributed-compute-dominodatalab-com-v1alpha1-raycluster,mutating=false,failurePolicy=fail,sideEffects=None,groups=distributed-compute.dominodatalab.com,resources=rayclusters,verbs=create;update,versions=v1alpha1,name=vraycluster.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Validator = &RayCluster{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *RayCluster) ValidateCreate() error {
+	logger.WithValues("raycluster", client.ObjectKeyFromObject(r)).Info("validating create")
+
+	return r.validateRayCluster()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (r *RayCluster) ValidateUpdate(old runtime.Object) error {
+	logger.WithValues("raycluster", client.ObjectKeyFromObject(r)).Info("validating update")
+
+	return r.validateRayCluster()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+// Not used, just here for interface compliance.
+func (r *RayCluster) ValidateDelete() error {
+	return nil
+}
+
+func (r *RayCluster) validateRayCluster() error {
+	var allErrs field.ErrorList
+
+	if err := r.validateWorkerReplicas(); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	if err := r.validateWorkerResourceRequestsCPU(); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	if err := r.validateObjectStoreMemoryBytes(); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	if errs := r.validatePorts(); errs != nil {
+		allErrs = append(allErrs, errs...)
+	}
+	if errs := r.validateAutoscaler(); errs != nil {
+		allErrs = append(allErrs, errs...)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "distributed-compute.dominodatalab.com", Kind: "RayCluster"},
+		r.Name,
+		allErrs,
+	)
+}
+
+func (r *RayCluster) validateWorkerReplicas() *field.Error {
+	replicas := r.Spec.Worker.Replicas
+	if replicas == nil || *replicas >= 0 {
+		return nil
+	}
+
+	return field.Invalid(
+		field.NewPath("spec").Child("worker").Child("replicas"),
+		replicas,
+		"should be greater than or equal to 0",
```

Comment:
Is this not potentially a desirable feature? E.g. could we support scale-down when not in use?

Reply:
I'm not sure I understand the first question. The predicate preceding this field error allows values `>= 0` so that users can scale the deployment to zero. If the HPA is present, it sees this as an [implicit maintenance-mode](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#implicit-maintenance-mode-deactivation). Am I missing something here?

Re: scale-down support, I'm going to assume you mean "HPA scale down to zero". According to the [docs](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates), this is only possible when the alpha feature gate `HPAScaleToZero` is enabled.

Reply:
uhyup, I misread this entirely.